light-client-verifier/src/operations/voting_power.rs

```diff
@@ -77,28 +112,37 @@ pub trait VotingPowerCalculator: Send + Sync {
         untrusted_validators: &ValidatorSet,
     ) -> Result<(), VerificationError> {
         let trust_threshold = TrustThreshold::TWO_THIRDS;
-        let voting_power =
-            self.voting_power_in(untrusted_header, untrusted_validators, trust_threshold)?;
-
-        if trust_threshold.is_enough_power(voting_power.tallied, voting_power.total) {
-            Ok(())
-        } else {
-            Err(VerificationError::insufficient_signers_overlap(
-                voting_power,
-            ))
-        }
+        self.voting_power_in(untrusted_header, untrusted_validators, trust_threshold)?
+            .check()
+            .map_err(VerificationError::insufficient_signers_overlap)
     }
 
-    /// Compute the voting power in a header and its commit against a validator set.
+    /// Compute the voting power in a header and its commit against a validator
+    /// set.
+    ///
+    /// Note that the returned tally may be lower than actual tally so long as
+    /// it meets the `trust_threshold`.
     ///
-    /// The `trust_threshold` is currently not used, but might be in the future
-    /// for optimization purposes.
+    /// If you have two separate sets of validators and need to check voting
+    /// power for both of them, prefer [`Self::voting_power_in_sets`] method.
     fn voting_power_in(
         &self,
         signed_header: &SignedHeader,
         validator_set: &ValidatorSet,
         trust_threshold: TrustThreshold,
     ) -> Result<VotingPowerTally, VerificationError>;
+
+    /// Compute the voting power in a header and its commit against two separate
+    /// validator sets.
+    ///
+    /// This is equivalent to calling [`Self::voting_power_in`] on each set
+    /// separately but may be more optimised.
+    fn voting_power_in_sets(
```

Comment:
Do we really need to expose this method in the trait, or can we instead just inline its implementation in `check_enough_trust_and_signers`?

Reply:
We don’t have to expose it. I guess it just feels natural to that the two required methods are voting_power_in and voting_power_in_sets.  I guess alternative is to get rid of those two methods and just leave `check_enough_trust_and_signers` and `check_signers_overlap` methods as required.

Reply:
Changed to remove vote_power_in and vote_power_in_sets. What do you think?

Reply:
You are right, after looking at the resulting code, it indeed makes more sense to keep `voting_power_in` and `voting_power_in_sets`. Because the latter is a bit ad-hoc, I was hoping we could do without it in favor of a more generic method like

```rust
trait VotingPowerCalculator {
   // ...

    fn compute_voting_power<V: signature::Verifier>(
        votes: &mut NonAbsentCommitVotes,
        validator_set: &ValidatorSet,
        trust_threshold: TrustThreshold,
        total_voting_power: u64,
    ) -> Result<VotingPowerTally, VerificationError>;
}
```

but that would require us to expose the `NonAbsentCommitVotes` struct publicly and add a `V: signature::Verifier` type parameter either to the `VotingPowerCalculator` trait or to each method, and I'd rather keep that as an implementation detail to allow other means of verifying signatures.

Let's thus revert to the previous commit.

What do you think about adding a default implementation for `voting_power_in_sets` that calls `voting_power_in` twice?

Reply:
> What do you think about adding a default implementation for voting_power_in_sets that calls voting_power_in twice?

I wondered about this and I think being a bit user-hostile here is justifiable.  Not having default implementation forces the implementator to think about the optimisation rather than using the old unoptimised code.  To ease the nagging equivalent code could be put in docstring so lazy implementators/people in a hurry could just copy that.

Reply:
Makes sense!

Reply:
I think this is ready for review then.